src/parcels/_core/sgrid.py

```diff
@@ -0,0 +1,373 @@
+            if not (
+                isinstance(vertical_dimensions, tuple)
+                and len(vertical_dimensions) == 1
+                and isinstance(vertical_dimensions[0], DimDimPadding)
+            ):
+                raise ValueError("vertical_dimensions must be a tuple of 1 DimDimPadding for a 2D grid")
+
+        # Required attributes
+        self.cf_role = cf_role
+        self.topology_dimension = topology_dimension
+        self.node_dimensions = node_dimensions
+        self.face_dimensions = face_dimensions
+
+        #! Optional attributes aren't really important to Parcels, can be added later if needed
+        # Optional attributes
+        # # With defaults (set in init)
+        # edge1_dimensions: tuple[Dim, DimDimPadding]
+        # edge2_dimensions: tuple[DimDimPadding, Dim]
+
+        # # Without defaults
+        # node_coordinates: None | Any = None
+        # edge1_coordinates: None | Any = None
+        # edge2_coordinates: None | Any = None
+        # face_coordinate: None | Any = None
+
+        #! Important optional attribute for 2D grids with vertical layering
+        self.vertical_dimensions = vertical_dimensions
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Grid2DMetadata):
+            return NotImplemented
+        return (
+            self.cf_role == other.cf_role
+            and self.topology_dimension == other.topology_dimension
+            and self.node_dimensions == other.node_dimensions
+            and self.face_dimensions == other.face_dimensions
+            and self.vertical_dimensions == other.vertical_dimensions
+        )
+
+    @classmethod
+    def from_attrs(cls, attrs):
+        try:
+            return cls(
+                cf_role=attrs["cf_role"],
+                topology_dimension=attrs["topology_dimension"],
+                node_dimensions=load_mappings(attrs["node_dimensions"]),
+                face_dimensions=load_mappings(attrs["face_dimensions"]),
+                vertical_dimensions=maybe_load_mappings(attrs.get("vertical_dimensions")),
+            )
+        except Exception as e:
+            raise SGridParsingException(f"Failed to parse Grid2DMetadata from {attrs=!r}") from e
+
+    def to_attrs(self) -> dict[str, str | int]:
+        d = dict(
+            cf_role=self.cf_role,
+            topology_dimension=self.topology_dimension,
+            node_dimensions=dump_mappings(self.node_dimensions),
+            face_dimensions=dump_mappings(self.face_dimensions),
+        )
+        if self.vertical_dimensions is not None:
+            d["vertical_dimensions"] = dump_mappings(self.vertical_dimensions)
+        return d
+
+
+class Grid3DMetadata(SGridMetadataProtocol):
+    def __init__(
+        self,
+        cf_role: Literal["grid_topology"],
+        topology_dimension: Literal[3],
+        node_dimensions: tuple[Dim, Dim, Dim],
+        volume_dimensions: tuple[DimDimPadding, DimDimPadding, DimDimPadding],
```

Comment:
Why si this `DimDimPadding` and not `DimDimDimpadding` as it is 3D here? Or `DimPadding`, as any number of dimensions is padded?